meetup.py

```diff
@@ -0,0 +1,69 @@
+"""Meetup API code and datastore model classes.
+"""
+import logging
+
+from granary import meetup as gr_meetup
+from oauth_dropins import meetup as oauth_meetup
+from oauth_dropins.webutil.util import json_dumps, json_loads
+import webapp2
+
+from models import Source
+import util
+
+# We don't support listen
+LISTEN_SCOPES = []
+PUBLISH_SCOPES = [
+  'rsvp',
+]
+
+
+class Meetup(Source):
+  GR_CLASS = gr_meetup.Meetup
+  OAUTH_START_HANDLER = oauth_meetup.StartHandler
+  SHORT_NAME = 'meetup'
+  BACKFEED_REQUIRES_SYNDICATION_LINK = True
+  CAN_PUBLISH = True
+  URL_CANONICALIZER = util.UrlCanonicalizer(
+    domain=GR_CLASS.DOMAIN,
+    headers=util.REQUEST_HEADERS)
+
+  @staticmethod
+  def new(handler, auth_entity=None, **kwargs):
+    """Creates and returns a :class:`Meetup` for the logged in user.
+
+    Args:
+      handler: the current :class:`webapp2.RequestHandler`
+      auth_entity: :class:`oauth_dropins.meetup.MeetupAuth`
+      kwargs: property values
+    """
+    user = json_loads(auth_entity.user_json)
+    gr_source = gr_meetup.Meetup(access_token=auth_entity.access_token())
+    actor = gr_source.user_to_actor(user)
+    return Meetup(id=auth_entity.key.id(),
+                  auth_entity=auth_entity.key,
+                  name=actor.get('displayName'),
+                  picture=actor.get('image', {}).get('url'),
+                  url=actor.get('url'),
+                  **kwargs)
+
+  def silo_url(self):
+    """Returns the Meetup account URL, e.g. https://meetup.com/members/...."""
+    return self.gr_source.user_url(self.key.id())
+
+  def label_name(self):
+    """Returns the username."""
+    return self.name
+
+
+class AddMeetup(oauth_meetup.CallbackHandler, util.Handler):
+  def finish(self, auth_entity, state=None):
+    logging.debug('finish with %s, %s', auth_entity, state)
+    self.maybe_add_or_delete_source(Meetup, auth_entity, state)
+
+
+ROUTES = [
+  ('/meetup/add', AddMeetup),
+  ('/meetup/start', oauth_meetup.StartHandler.to(
```

Comment:
ah, i think you need to add `util.oauth_starter(...` here to set `feature`. eg:

https://github.com/snarfed/bridgy/blob/70cca8fa98217f33e6f0abed06727a17ffcea1b7/github.py#L101-L102

Reply:
Good spot. That now returns the error:

```
HTTP Error 400: 400 Bad Request The server could not comply with the request since it is either malformed or otherwise incorrect. Invalid state value '{"feature":"publish","operation":"add"}' 
```

when returning the code :thinking:

Reply:
This is coming from oauth-dropins https://github.com/snarfed/oauth-dropins/blob/bbc62c2e064b789db581e179fa6204a3020b67be/oauth_dropins/meetup.py#L128-L133 - I'll look at if we've done something different for i.e. GitHub.

Reply:
Ah, I think this may be because it's the first of the OAuth consumers that we're using this with, so it just may not work at all. I'll temporarily remove the check to see if that unblocks this.

Reply:
Partially fixed by https://github.com/snarfed/oauth-dropins/pull/27 - but still investigating the other changes required. It's definitely something going wrong with the state parameter!

Reply:
^ That PR resolves the issue completely with CSRFs!